File: src/test/fixtures/python/cognitive_rules.py
# straightLine: cognitive score 0.
def straightLine(ready, enabled, urgent):
    return ready

# booleanRun: cognitive score 1.
def booleanRun(ready, enabled, urgent):
    return ready and enabled and urgent

# mixedRuns: cognitive score 3.
def mixedRuns(ready, enabled, urgent):
    return ready and enabled or urgent and ready

# groupedRun: cognitive score 1.
def groupedRun(ready, enabled, urgent):
    return ready and (enabled and urgent)

# negatedGroup: cognitive score 2.
def negatedGroup(ready, enabled, urgent):
    return ready and not (enabled and urgent)

# plainElse: cognitive score 2.
def plainElse(ready, enabled, urgent):
    if ready:
        deliver()
    else:
        queue()

# flatChain: cognitive score 3.
def flatChain(ready, enabled, urgent):
    if ready:
        deliver()
    elif urgent:
        expedite()
    else:
        queue()

# nestedChain: cognitive score 4.
def nestedChain(ready, enabled, urgent):
    if enabled:
        if ready:
            deliver()
        elif urgent:
            expedite()

# nestedChecks: cognitive score 3.
def nestedChecks(ready, enabled, urgent):
    if enabled:
        if ready:
            deliver()

# dispatch: cognitive score 3.
def dispatch(ready, enabled, urgent):
    match ready:
        case True:
            if enabled: deliver()
        case False:
            queue()

# loopBody: cognitive score 3.
def loopBody(ready, enabled, urgent):
    while ready:
        if enabled: deliver()

# lambdaBody: cognitive score 2.
def lambdaBody(ready, enabled, urgent):
    action = lambda: deliver() if ready else queue()
    action()

# emptyLambda: cognitive score 0.
def emptyLambda(ready, enabled, urgent):
    action = lambda: deliver()
    action()

# guardReturn: cognitive score 1.
def guardReturn(ready, enabled, urgent):
    if not ready: return
    deliver()

# catches: cognitive score 4.
def catches(ready, enabled, urgent):
    try:
        deliver()
    except TimeoutError:
        if ready: queue()
    except ValueError:
        reject()

# cleanup: cognitive score 0.
def cleanup(ready, enabled, urgent):
    try:
        deliver()
    finally:
        cleanup()

# nestedFunction: cognitive score 2.
def nestedFunction(ready, enabled, urgent):
    audit()
    def action():
        if ready: deliver()
    action()

# emptyFunction: cognitive score 0.
def emptyFunction(ready, enabled, urgent):
    audit()
    def action():
        deliver()
    action()

# bracedElseCheck: cognitive score 4.
def bracedElseCheck(ready, enabled, urgent):
    if enabled:
        deliver()
    else:
        if ready:
            deliver()

File: src/test/fixtures/fsharp/CognitiveRules.fs
module CognitiveRules

// straightLine: cognitive score 0.
let straightLine (ready: bool) (enabled: bool) (urgent: bool) = ready

// booleanRun: cognitive score 1.
let booleanRun (ready: bool) (enabled: bool) (urgent: bool) = ready && enabled && urgent

// mixedRuns: cognitive score 3.
let mixedRuns (ready: bool) (enabled: bool) (urgent: bool) = ready && enabled || urgent && ready

// groupedRun: cognitive score 1.
let groupedRun (ready: bool) (enabled: bool) (urgent: bool) = ready && (enabled && urgent)

// negatedGroup: cognitive score 2.
let negatedGroup (ready: bool) (enabled: bool) (urgent: bool) = ready && not (enabled && urgent)

// plainElse: cognitive score 2.
let plainElse (ready: bool) (enabled: bool) (urgent: bool) = if ready then deliver () else queue ()

// flatChain: cognitive score 3.
let flatChain (ready: bool) (enabled: bool) (urgent: bool) =
    if ready then deliver ()
    elif urgent then expedite ()
    else queue ()

// nestedChain: cognitive score 4.
let nestedChain (ready: bool) (enabled: bool) (urgent: bool) =
    if enabled then
        if ready then
            deliver ()
        elif urgent then
            expedite ()

// nestedChecks: cognitive score 3.
let nestedChecks (ready: bool) (enabled: bool) (urgent: bool) =
    if enabled then
        if ready then
            deliver ()

// dispatch: cognitive score 3.
let dispatch (ready: bool) (enabled: bool) (urgent: bool) =
    match ready with
    | true ->
        if enabled then
            deliver ()
    | false -> queue ()

// loopBody: cognitive score 3.
let loopBody (ready: bool) (enabled: bool) (urgent: bool) =
    while ready do
        if enabled then
            deliver ()

// lambdaBody: cognitive score 2.
let lambdaBody (ready: bool) (enabled: bool) (urgent: bool) =
    let action =
        fun () ->
            if ready then
                deliver ()

    action ()

// emptyLambda: cognitive score 0.
let emptyLambda (ready: bool) (enabled: bool) (urgent: bool) =
    let action = fun () -> deliver ()
    action ()

// unitConditional: cognitive score 1.
let unitConditional (ready: bool) (enabled: bool) (urgent: bool) =
    if not ready then
        ()

    deliver ()

// catches: cognitive score 4.
let catches (ready: bool) (enabled: bool) (urgent: bool) =
    try
        deliver ()
    with
    | :? System.TimeoutException ->
        if ready then
            queue ()
    | :? System.ArgumentException -> reject ()

// cleanup: cognitive score 0.
let cleanup (ready: bool) (enabled: bool) (urgent: bool) =
    try
        deliver ()
    finally
        cleanup ()

// nestedFunction: cognitive score 2.
let nestedFunction (ready: bool) (enabled: bool) (urgent: bool) =
    audit ()

    let action () =
        if ready then
            deliver ()

    action ()

// emptyFunction: cognitive score 0.
let emptyFunction (ready: bool) (enabled: bool) (urgent: bool) =
    audit ()
    let action () = deliver ()
    action ()

// bracedElseCheck: cognitive score 4.
let bracedElseCheck (ready: bool) (enabled: bool) (urgent: bool) =
    if enabled then
        deliver ()
    else
        (if ready then
             deliver ())
